Update deprecated back pressed event

irfano opened the pull request:
This PR is part of a [parent PR](https://github.com/wordpress-mobile/WordPress-Android/pull/17947) to update compileSdk to 33.

`onBackPressed()` was [deprecated on Android 13](https://developer.android.com/guide/navigation/navigation-custom-back). This updates deprecated usages of onBackPressed.

- [Use onBackPressedDispatcher to trigger onBackPressed()](https://github.com/wordpress-mobile/WordPress-Android/commit/8648d0bbb9b7994c6ce6f783bfadd164abb9f585): Activity's `onBackPressed()` calls are removed and updated with `OnBackPressedDispatcher`s `onBackPressed()`.
- [Migrate activities to OnBackPressedCallback](https://github.com/wordpress-mobile/WordPress-Android/commit/3bca5a7943c0a1e51ef15ea981cae2b0fdb26ecd): Overridings Activity's `onBackPressed` are removed. Instead, a callback is added to `onBackPressedDispatcher`. When I want to trigger `super.onBackPressed()`, I disabled the callback and triggered `OnBackPressedDispatcher`'s `onBackPressed()` gain.
```
onBackPressedDispatcher.addCallback(this) {
     ...
     // The code below disables this callback and lets the system handle the back pressed (e.g., by finishing the activity)
     isEnabled = false
     onBackPressedDispatcher.onBackPressed()
}
```
- [Migrate dialogs to OnBackPressedCallback](https://github.com/wordpress-mobile/WordPress-Android/commit/bf693ec6077d0a708982a158e9361fc4ad125df7): We need to cast dialogs to `ComponentDialog` to be able to use `OnBackPressedDispatcher`.  For this purpose, `Dialog(requireContext(), theme)` is converted to `super.onCreateDialog(savedInstanceState)`

To test:
Make sure the project is buildable and test back navigation on screens. It's very time-consuming to test all screens. But please test at least some screens from different cases, like a screen from an activity and a DialogFragment.

## Regression Notes
1. Potential unintended areas of impact
The back navigation function could be broken on some screens.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I haven't tested all screens, but I tested some activities and all `DialogFragment`s.

3. What automated tests I added (or what prevented me from doing so)
No new feature was added, and the existing tests were relied upon.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

irfano committed:
Use onBackPressedDispatcher to trigger onBackPressed()

 Activity's onBackPressed was deprecated on Android 13.

irfano committed:
Migrate activities to OnBackPressedCallback

Activity's onBackPressed was deprecated on Android 13.

irfano committed:
Use extension function for OnBackPressedCallback

This just makes the code shorter by using an extension function.

irfano committed:
Move lambda argument out pf parentheses

irfano committed:
Remove redundant null checks

irfano committed:
Migrate dialogs to OnBackPressedCallback

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailActivity.kt:
```diff
@@ -19,6 +20,9 @@ class HistoryDetailActivity : LocaleAwareActivity() {
             setSupportActionBar(toolbarMain)
         }
+
+        onBackPressedDispatcher.addCallback(this) { AnalyticsTracker.track(Stat.REVISIONS_DETAIL_CANCELLED) }
```
Warning (⚠️): As per the other such changes you did to replicate the `super.onBackPressed()` behavior, it seems that the `isEnabled = false` + `onBackPressedDispatcher.onBackPressed()` additions are missing from this callback.

PS: Same applies on the `SuggestionActivity.kt` class and its `onBackPressedDispatcher.addCallback(this) { viewModel.trackExit(false) }` replacement, which too seems to be missing the above.

irfano replied:
Good catch! I added the missing ones: cc9f2e9b0570f21c1bb4d4dd6ba2b159a8cdcb31

ParaskP7 replied:
Awesome, thanks for the fix @irfano ! 🥇

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt:
```diff
@@ -110,52 +113,47 @@ class SiteCreationActivity : LocaleAwareActivity(),
-            super.onBackPressed()
-        })
-        siteCreationIntentsViewModel.onBackButtonPressed.observe(this, Observer {
+            onBackPressedDispatcher.onBackPressed()
```
Question (❓): This `onBackPressedDispatcher.onBackPressed()` is replacing a `super.onBackPressed()` call and not a normal `onBackPressed()` call, is that intended here? 🤔

irfano replied:
Another good catch! The behavior was also broken here.
Fixed: 3a3110e41098bce85f6d8cfd900ef481c7801279

ParaskP7 replied:
Awesome, thanks for the fix @irfano ! 🥇 

Btw, this is the only place do also `backPressedCallback.isEnabled = true` afterwards, was there a reason for that? 🤔

irfano replied:
> Btw, this is the only place do also backPressedCallback.isEnabled = true afterwards, was there a reason for that? 🤔

`super.onBackPressed()` closes the fragment while `SiteCreationActivity`  is still alive. That's why we still need the callback to be enabled after calling `onBackPressedDispatcher.onBackPressed()`. This is an example of the concern I shared in the [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/18019#discussion_r1123718077). 
Other classes don't have the same behavior, they just finish the activity on `super.onBackPressed()`. But still, to be on the safe side, I have decided to do the same everywhere, not just here. Additionally, I will be using an extension function for better manageability.

ParaskP7 replied:
Great, thanks for the explanation and for applying the same behavior everywhere else @irfano ! 🚀

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jpfullplugininstall/onboarding/JetpackFullPluginInstallOnboardingDialogFragment.kt:
```diff
@@ -59,12 +61,13 @@ class JetpackFullPluginInstallOnboardingDialogFragment : DialogFragment() {
                 viewModel.onDismissScreenClick()
-                super.onBackPressed()
+                isEnabled = false
+                componentDialog.onBackPressedDispatcher.onBackPressed()
```
Suggestion (💡): Consider updating this to the equivalent change you did on `FeatureIntroductionDialogFragment`, see below:

```
(this as ComponentDialog).onBackPressedDispatcher.addCallback(this@JetpackFullPluginInstallOnboardingDialogFragment) {
    viewModel.onDismissScreenClick()
    isEnabled = false
    onBackPressedDispatcher.onBackPressed()
}
```

irfano replied:
There was also another issue here that caused a crash. I tried to cast `dialog` to `ComponentDialog`, but `dialog` is null in `onCreateDialog()` since it hasn't been created yet.
Fixed: 1808fd4a471d71b87f598ddbc4c4e6f9aa2d64dc

ParaskP7 replied:
Nice, thanks for the fix @irfano ! 🥇

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/CollapseFullScreenDialogFragment.java:
```diff
@@ -146,12 +148,14 @@ public void setConfirmEnabled(boolean enabled) {
     public Dialog onCreateDialog(Bundle savedInstanceState) {
         initBuilderArguments();
 
-        Dialog dialog = new Dialog(requireContext(), getTheme()) {
```
Minor (🔍): Consider making the `(ComponentDialog)` cast here instead of doing that further below just to get access to `getOnBackPressedDispatcher()`.

```
ComponentDialog dialog = (ComponentDialog) super.onCreateDialog(savedInstanceState);
```

PS: Same applies for the `FullScreenDialogFragment` change.

irfano replied:
Done: 1749865ed8a954c7dda33e64f2ea430cf5e93458

ParaskP7 replied:
Many thanks for this change @irfano ! 🥇

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentsDetailActivity.java:
```diff
@@ -94,6 +83,22 @@ public void onCreate(Bundle savedInstanceState) {
+                        getSupportFragmentManager().findFragmentByTag(CollapseFullScreenDialogFragment.TAG);
+
+                if (fragment != null) {
+                    fragment.onBackPressed();
```
Minor (🔍): Consider renaming this publicly available, but not `@Override`, method to something else. It can become misleading and others might incorrectly assume that this is indeed an `@Override onBackPressed` method that needs to be used as such from all these 3 activities that are calling it, I know because I did just that at first... 😅

Maybe a better name for that method would be something like `checkAddedAndCollapse()`? 🤔 

PS: Same applies to `FullScreenDialogFragment`, but this method there seems to be unused as well... 🤷

irfano replied:
I was also confused and checked if `onBackPressed` is an override function. 😄 It will be a good action to rename it. 
What about renaming it to `collapse()`? The `checkAdded...` part doesn't seem to have much meaning to me.

I believe we can remove `onBackPressed` from `FullScreenDialogFragment` as it is currently unused. I see no reason not to remove it.

ParaskP7 replied:
👋 @irfano !

> What about renaming it to collapse()? The checkAdded... part doesn't seem to have much meaning to me.

Yea, let's better rename it to something as simple as `collapse()`. 👍 💯 

> I believe we can remove onBackPressed from FullScreenDialogFragment as it is currently unused. I see no reason not to remove it.

Coolio, let's remove it then, I agree! 👍 💯

irfano replied:
Done: fb7a69c9b73653838f56ff7d1bd3ffc3de2bbb56

ParaskP7 replied:
Awesome! 🚀

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/deeplinks/DeepLinkingIntentReceiverActivity.java:
```diff
@@ -49,6 +50,17 @@ public class DeepLinkingIntentReceiverActivity extends LocaleAwareActivity {
+            public void handleOnBackPressed() {
+                setEnabled(false);
+                getOnBackPressedDispatcher().onBackPressed();
+                finish();
```
Question (❓): I was trying to figure out what's the best way to deal with both, replacing these `super.onBackPressed()` calls and the `finish()` calls, thus I have some question about both, see below:

1. `super.onBackPressed()`: Is `setEnabled(false)` + `getOnBackPressedDispatcher().onBackPressed()` the recommended way to migrate those calls when utilizing `OnBackPressedCallback`. Do you maybe have some references on that so I could take a quick look? Unfortunately, I couldn't find any references on using this pattern myself. PS: No matter, from my quick testing, this indeed seems to be doing the trick. I am just not sure if that't the best we can do atm, especially after taking a look at [this](https://www.droidcon.com/2023/02/22/handling-back-press-in-android-13-the-correct-way/).
2. `finish()`: After me taking a look at [this](https://www.droidcon.com/2023/02/22/handling-back-press-in-android-13-the-correct-way/), the author mentioned that adding such calls will break the predictive back gesture feature. I know, we don't use `android:enableOnBackInvokedCallback="true"` atm, but that made me wonder if we can do better anyway, and avoid redoing that later on, wdyt? 🤔 

FYI: About no.2 above, in addition to this class and after me doing a quick search, I found the below additional activities to be using such `finish()` call:
- `JetpackSecuritySettingsActivity` via `finish()`
- `PostSettingsTagsActivity` via `saveAndFinish()`
- `SelectCategoriesActivity` via `saveAndFinish()`
- `JetpackConnectionResultActivity` via `finishAndGoBackToSource()`

irfano replied:
Thank you for initiating this discussion. I have spent a lot of time thinking about how to migrate `super.onBackPressed()`. Unfortunately, I couldn't find any official recommendations for migrating `super.onBackPressed()`. However, did come across some [references](https://issuetracker.google.com/issues/247982487) that do not recommend the current solution here if the predictive gesture feature is enabled.
I believe that the solution proposed here simply mimics the behavior of `super.onBackPressed()`, but it will have side effects when `enableOnBackInvokedCallback` is set to `true`. Therefore, it appears safe to me, but I would appreciate a review of my thought process.

---
**The problem**
When the back button is pressed,
- If an enabled `OnBackPressedCallback` exists, the system does nothing. It does not finish the activity, but only calls the callback function.
- If no enabled `OnBackPressedCallback` exists, the system handles the back button and finishes the activity.

We want the system to handle the back button, but we also want to run our custom code before the system. 

**The current solution**
I looked for a more elegant solution, but unfortunately, I could not find one. Therefore, I disabled the `OnBackPressedCallback` (`setEnabled(false)`) and then called `getOnBackPressedDispatcher().onBackPressed()` again to allow the system to handle the back button and keep the behavior of `super.onBackPressed()`. 

**A small concern for the current solution**
My only concern is that we may still need the callback after calling the `getOnBackPressedDispatcher().onBackPressed()`, but the callback will be disabled in this case. If `getOnBackPressedDispatcher().onBackPressed()` finishes the activity, there will be no issue. However, the system may just [collapse the drawer](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/Activity.java#3958) with the back button. I have checked our cases and could not find any case where this issue occurred. We can still add `setEnabled(true)` after `getOnBackPressedDispatcher().onBackPressed()` just in case, but I am of the opinion that it is not necessary. **Wdyt?**.

---

>  adding such calls will break the predictive back gesture feature. I know, we don't use android:enableOnBackInvokedCallback="true" atm, but that made me wonder if we can do better anyway, and avoid redoing that later on, wdyt? 🤔

`finish()` is breaking the predictive back feature, and `setEnabled(false) + getOnBackPressedDispatcher().onBackPressed()` is also [breaking](https://issuetracker.google.com/issues/247982487#comment3) it. Therefore, if I don't miss anything, there's no way to use custom back navigation and finish your activity. 😅
I'm aware of the risk of redoing some work here. However,  I prefer to keep the `predictive back gesture` out of [updating compileSdk issue](https://github.com/wordpress-mobile/WordPress-Android/pull/17947)'s scope. My other reasons for not enabling it are that the predictive gesture feature also needs upgrading to `androidx.activity:activity:1.6.0`, and it is a feature for Android 14. We can consider predictive back gesture as a separate project. 

---
**Summary:** I prefer to keep the current implementation if you cannot find any side effects of it like me.

ParaskP7 replied:
👋 @irfano !

> Thank you for initiating this discussion. I have spent a lot of time thinking about how to migrate super.onBackPressed(). Unfortunately, I couldn't find any official recommendations for migrating super.onBackPressed(). However, did come across some [references](https://issuetracker.google.com/issues/247982487) that do not recommend the current solution here if the predictive gesture feature is enabled.

Great find on this [references](https://issuetracker.google.com/issues/247982487) that also don't recommend the current solution. 👍 

> I believe that the solution proposed here simply mimics the behavior of super.onBackPressed(), but it will have side effects when enableOnBackInvokedCallback is set to true. Therefore, it appears safe to me, but I would appreciate a review of my thought process.

It indeed seems safe until we enable the predictive back gesture, that was my only concern. 👍 

> The problem

👍 

> The current solution

👍 

> A small concern for the current solution

I have the exact same concern, but I was also on the fence about adding this `setEnabled(true)` after `getOnBackPressedDispatcher().onBackPressed()`. But maybe we should, just to be on the safe side, not sure... 🤔

Suggestion (💡): An additional thought I had on that, just because of the fact that we keep replicating this behavior in multiple places, what if we create a few extension functions, like the below `AppCompatActivity` related one, and use them instead:

`AppCompatActivity`
```
/**
 * This is a temporary workaround for the issue described here: https://issuetracker.google.com/issues/247982487
 *
 * TODO: Replace this temporary workaround before enabling predictive back gesture on this project
 *  (android:enableOnBackInvokedCallback="true").
 *
 * Related Issue: <LINK>
 */
fun AppCompatActivity.onBackPressedCompat(onBackPressedCallback: OnBackPressedCallback) {
    onBackPressedCallback.isEnabled = false
    onBackPressedDispatcher.onBackPressed()
    onBackPressedCallback.isEnabled = true
}
```

> finish() is breaking the predictive back feature, and setEnabled(false) + getOnBackPressedDispatcher().onBackPressed() is also [breaking](https://issuetracker.google.com/issues/247982487#comment3) it. Therefore, if I don't miss anything, there's no way to use custom back navigation and finish your activity. 😅

😅 Yea, I too am not sure about it atm, I would have to research that a bit more to figure out if there is a way to use a custom back navigation and finish your activity. 🤷 

> I'm aware of the risk of redoing some work here. However, I prefer to keep the predictive back gesture out of https://github.com/wordpress-mobile/WordPress-Android/pull/17947's scope. My other reasons for not enabling it are that the predictive gesture feature also needs upgrading to androidx.activity:activity:1.6.0, and it is a feature for Android 14. We can consider predictive back gesture as a separate project.

For sure, my intention with this comment of mine wasn't about us enabling predictive back gesture here, not at all, I would actually advice against that, this is a whole other work that we would only be able to do in the future (as you noted too). 💯 

However, now that we are at it, I would advise us to create that future predictive back gesture issue, noting all these details and reference this issue on the KDoc comments that we might add as part of the extension functions idea that I suggested above. Wdyt?

> Summary: I prefer to keep the current implementation if you cannot find any side effects of it like me.

👍

irfano replied:
> Suggestion (💡): An additional thought I had on that, just because of the fact that we keep replicating this behavior in multiple places, what if we create a few extension functions

I was just about to do that before reading your comment. It's good that we're thinking alike. Done: bfcf01d2e320bba9894ace5677b6d43b865584f5

I opened the [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/18053) as you suggested. I also received a [response](https://stackoverflow.com/questions/75462234/proper-usage-of-onbackpresseddispatcher/75462987#comment133451000_75462987) from a Google Engineer. According to the answer, we should update all of our back-pressed approaches. Since the purpose of custom back navigation differs for each class, the solution should also be different. Still, I'm for leaving that change to the other [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/18053).

ParaskP7 replied:
This is awesome @irfano and thank you so much for going through this change, it does feel better indeed! 🙇  ❤️ 🚀

ParaskP7 reviewed (changes requested):
👋 @irfano !

Thank you so much for opening this PR, the work you did here really take us to the next level. This work will (hopefully) help us avoid redoing any such `onBackPressed` work in the future, and, no more dealing with deprecation such as that, win-win! 🎉 

I reviewed and quickly tested your changes, I did identify some problems (I think, see ⚠️) and also posted few questions that I would like to discuss with you (see ❓), that is, before us merging this to its parent branch.

Thus, at this point I am going to request changes, but don't be alarmed too much. After we resolve/discuss on them, I am going to do another round of testing to verify that everything works as expected and we'll then merge this to its parent branch asap. 👍

irfano committed:
Add missing super.onBackPressed() behaviors

irfano committed:
Fix the crash on JetpackFullPluginInstallOnboardingDialogFragment

irfano committed:
Fix loop issue of OnBackPressedCallback in SiteCreationActivity

OnBackPressedCallback was temporarily disabled to allow the system to handle the back button event.

irfano committed:
Refactor ComponentDialog casting

irfano committed:
Refactor onBackPressed function in CollapseFullScreenDialogFragment

irfano committed:
Use CompatExtensions functions for migrating super.onBackPressed

ParaskP7 reviewed (approved):
👋 @irfano and once more, thank you so much for all the work you did with this PR, research and coding included! 🥇 

After reviewing the update, I now feel much more comfortable hitting the ✅ button! 🎉 

> Thus, at this point I am going to request changes, but don't be alarmed too much. After we resolve/discuss on them, I am going to do another round of testing to verify that everything works as expected and we'll then merge this to its parent branch asap. 👍

I did another round of testing and can verify that everything worked as expected, I didn't at least notice a crash... 💯 

But, be mindful that I didn't go through all the screens/functionality to verify every back pressed logic. As such, and as discussed/planned already, I recommend that when the parent [feature/update-compile-sdk-33](https://github.com/wordpress-mobile/WordPress-Android/tree/feature/update-compile-sdk-33) gets ready we take it for another spin of testing, this time thoroughly, while asking from QE to help us on that too.

Let's merge this! 🚀

irfano commented:
> But, be mindful that I didn't go through all the screens/functionality to verify every back pressed logic. As such, and as discussed/planned already, I recommend that when the parent [feature/update-compile-sdk-33](https://github.com/wordpress-mobile/WordPress-Android/tree/feature/update-compile-sdk-33) gets ready we take it for another spin of testing, this time thoroughly, while asking from QE to help us on that too.

I am about to submit a request for a QE. I'll mention that back navigation is an important change that needs to be tested.
Thank you for your thorough review! You can merge it as the reviewer. 🚀
